streams/src/test/java/org/apache/kafka/test/KStreamTestDriver.java

```diff
@@ -120,6 +120,22 @@ public StateStore getStateStore(String name) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    public <K, V> void forward(K key, V value, String childName) {
+        ProcessorNode thisNode = currNode;
+        try {
+            for (ProcessorNode childNode : (List<ProcessorNode<K, V>>) thisNode.children()) {
+                if (childNode.name().equals(childName)) {
+                    currNode = childNode;
```

Comment:
This is pervasive here and also in StreamTask.java. My understanding is that a task passes/forwards a record through multiple ProcessorNodes, and each of them calls process(). process() in turn could go depth-first in the ProcessorNode graph. As each node is processed (depth-first), the top caller restores the starting node. @guozhangwang does this match your understanding?